### PR TITLE
Add grouped listbox example

### LINKED ISSUE #913: Listbox example: Update to illustrate use of group role for grouping options
Resolution of w3c/aria#719 changes ARIA 1.2 to allow option elements to be contained in an element with role group. Update a listbox example to demonstrate option grouping by:

- [ ] Adding a labeled group to the options
- [ ] Documenting the labeled group in the roles, states, and properties table on the example page.
- [ ] Updating the listbox tests.

This is related to issue #765.

### PR DESCRIPTION
**Pull request by smhigley**
Adds a listbox pattern with grouped options with tests.

Additionally, fixes these bugs with clicking listbox options:
- Clicking any option below the initially visible aria results in the first option being selected instead of the clicked option
- Clicking any option within the initially visible area briefly selects the first option before moving selection to the clicked option

By:
- not immediately selecting the first option when the listbox receives focus (bringing it more inline with native `<select multiple>`
- only initiating automatic listbox scroll when interacting with the keyboard

Resolves #913 

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @ZoeBijl @smhigley 
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @ZoeBijl 
* [x] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) @ZoeBijl @carmacleod 
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @ZoeBijl 
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by @spectranaut 
### Preview link

[View this example in the compare branch for this PR](https://raw.githack.com/w3c/aria-practices/913-listbox-groups/examples/listbox/listbox-grouped.html)

### CODE REVIEW

File: examples/listbox/listbox-grouped.html [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+      <li>
+        Scrolling only works as expected if the listbox is the options' <code>offsetParent</code>.
+        The example uses <code>position: relative</code> on the listbox to that effect.
+      </li>
+      <li>When an option is focused that isn't (fully) visible, the listbox's scroll position is updated:
+        <ol>
+          <li>If <kbd>Up Arrow</kbd> or <kbd>Down Arrow</kbd> is pressed, the previous or next option is scrolled into view.</li>
+          <li>If <kbd>Home</kbd> or <kbd>End</kbd> is pressed, the listbox scrolls all the way to the top or to the bottom.</li>
+          <li>If <code>focusItem</code> is called, the focused option will be scrolled to the top of the view if it was located above it or to the bottom if it was below it.</li>
+          <li>If the mouse is clicked on a partially visible option, it will be scrolled fully into view.</li>
```
Comment (ZoeBijl) [inline examples/listbox/listbox-grouped.html]: This isn’t happening for me in either latest stable Chrome or Safari on macOS 10.16.

File: examples/listbox/listbox-grouped.html [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      <li><a href="https://github.com/w3c/aria-practices/issues/new">Report Issue</a></li>
+      <li><a href="https://github.com/w3c/aria-practices/projects/13">Related Issues</a></li>
+      <li><a href="../../#Listbox">Design Pattern</a></li>
+    </ul>
+  </nav>
+  <main>
+  <h1>Listbox Example with Grouped Options</h1>
+  <p>
+    The following example implementation of the <a href="../../#Listbox">design pattern for listbox</a> demonstrates a single-select listbox widget with grouped options.
+    This widget is functionally similar to an HTML <code>select</code> element with labelled <code>optgroup</code> elements used to group options.
```
Comment (ZoeBijl) [inline examples/listbox/listbox-grouped.html]: Would it be worth to include something like “…similar to an HTML <code>select</code> element with the `multiple` attribute set that contains labelled <code>optgroup</code> elements…”? But perhaps with less awkward wording than my pre coffee comment. Not sure if this is necessary tho.

File: examples/listbox/css/listbox.css [hunk 3]
```diff
@@ -34,6 +34,20 @@
   border: 1px solid #aaa;
 }
 
+[role="group"] {
+  padding: 0;
+  margin: 0;
+}
+
+[role="group"] > [role="presentation"] {
```
Comment (ZoeBijl) [inline examples/listbox/css/listbox.css]: Yas to all the attribute selectors 👏

File: examples/listbox/listbox-grouped.html (hunk 2 above)
Comment (smhigley) [reply examples/listbox/listbox-grouped.html]: I think it's still only single-select; maybe "similar to an HTML `select` element with `size` > 1 ..." would work?

File: examples/listbox/listbox-grouped.html (hunk 1 above)
Comment (smhigley) [reply examples/listbox/listbox-grouped.html]: Pushing a fix + some tests for it in just a bit!

Comment (smhigley) [reply examples/listbox/listbox-grouped.html]: Updated, thanks for the catch!

File: examples/listbox/listbox-grouped.html (hunk 2 above)
Comment (ZoeBijl) [reply examples/listbox/listbox-grouped.html]: Ah yes of course. Thanks for the addition!

**Comment by ZoeBijl** [comment]
@mcking65 no, the current approval is from 17 days ago. At which point I was only assigned functional review. Is this the one you wanted to get done this week?

File: test/tests/listbox_grouped.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  t.true(listboxBounds.y + listboxBounds.height - optionBounds.y < 0, 'last option is not initially displayed');
+
+  await listbox.sendKeys(Key.END);
+  listboxBounds = await listbox.getRect();
+  optionBounds = await options[options.length - 1].getRect();
+
+  t.true(listboxBounds.y + listboxBounds.height - optionBounds.y >= 0, 'last option is in view after end key');
+});
+
+ariaTest('Click scrolls listbox option into view', exampleFile, 'key-end', async (t) => {
```
Comment (spectranaut) [inline test/tests/listbox_grouped.js]: I see your desire for a way to tests other attributes of the example :) Good test.

Comment (smhigley) [reply test/tests/listbox_grouped.js]: Thanks 😊

**Comment by smhigley** [comment]
Thanks for taking care of those editorial updates, @mcking65!

**Comment by ZoeBijl** [comment]
Visual design and code review done. I would appreciate it if someone were to check the JavaScript as I only glanced that. One thing I did notice was the @param etc comment stuff. Since this is adding to an existing pattern, we should either remove all of those from all listbox examples or leave them in for now.

Summary of what I did:
- Rearrange a bunch of CSS properties
- Simplify the CSS
  - Made the listbox area a grid
  - Used more attribute selectors instead of classes
  - Removed `font-size: 0` hacks
- Made vertical flow consistent
- Removed some unnecessary classes from the HTML

File: examples/listbox/listbox-grouped.html [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+              <li id="ss_elem_9" role="option">Falcon</li>
+              <li id="ss_elem_10" role="option">Winged Horse</li>
+              <li id="ss_elem_11" role="option">Owl</li>
+            </ul>
+          </div>
+        </div>
+      </div>
+    </div>
+    <div role="separator" id="ex_end_sep" aria-labelledby="ex_end_sep ex_label" aria-label="End of"></div>
+    <h4>Notes</h4>
```
Comment (carmacleod) [inline examples/listbox/listbox-grouped.html]: The "Notes" heading should be `<h3>` instead of `<h4>` to avoid skipping a level.

Comment (smhigley) [reply examples/listbox/listbox-grouped.html]: Fixed, thanks @carmacleod!

Comment (carmacleod) [reply examples/listbox/listbox-grouped.html]: Wow - thanks for fixing so quickly!

**Comment by carmacleod** [comment]
### Accessibility Testing Notes

#### Screen Reader/Browser Pairings Tested

JAWS | JAWS | JAWS | JAWS | JAWS |
------ | ------ | ------ | ------ | ------ |
 IE11 |  FF  | Chrome | Edge | Edge Beta |
  pass  | pass `1` | pass | fail `2` | pass |

 NVDA | NVDA | NVDA | NVDA | NVDA | 
 ------ | ------ | ------ | ------ | ------ |
 IE11 |  FF  | Chrome | Edge | Edge Beta |
 pass | pass `1` | pass | ok `3` | pass |

 Narrator | Narrator |
 ------ | ------ |
  Edge | Edge Beta |
 ok `4` `5` | ok `4` |

1. Firefox exposes the "checkmark" CSS content. JAWS reads this as "checkmark Dog selected", and NVDA reads it as "checked Dog".

2. JAWS in Edge (non-Chromium) doesn't work very well for this example. I had to manually enter focus mode in the listbox, but even then, arrow keys were behaving as though it was still in browse mode, and options had to be explicitly selected with space or enter.

3. NVDA in Edge (non-Chromium) doesn't say selected state, and doesn't say group name, but it recognizes groups exist and adjusts position and set size accordingly.

4. Narrator doesn't recognize the groupings in the listbox. In scan mode, arrow keys read the group name, but in focus mode, groups are ignored. I also noted that a listbox is called a "Selection" (from UIA Selection Control Pattern).

5. Had to manually enter focus mode in the listbox in non-Chromium Edge.

6. Will test VO on Mac OS and iOS tomorrow. I don't have access to an Android device or Chromebook. Don't know if we want to test with Dragon or ZoomText.

    Windows 10 Enterprise 1903
    JAWS 2019.1909.28
    NVDA 2019.2.1
    Version 78.0.3904.87
    Firefox 71.0b7
    Edge 44.18362.387.0
    Edge Beta 79.0.309.12

#### Automated Testing Tools

Played with lots of fun automated testing tools. :)
- aXe in Chrome DevTools
- WAVE
- Tota11y
- Accessibility Insights
- IBM DAP

All automated tests passed, except for a skipped h3, which @smhigley fixed right away. (thanks! :)

#### Windows High Contrast Mode
I was surprised/not surprised to note that Edge Beta (Chromium) doesn't support WHCM.
So only IE11 and Firefox needed to be tested in WHCM (and Edge... until Jan 15). All passed.

Listbox Example in Firefox in Windows High Contrast Mode:
screen snap

**Comment by carmacleod** [comment]
#### Screen Reader/Browser Pairings Tested (cont'd)

VoiceOver | VoiceOver | VoiceOver |
------ | ------ | ------ |
 macOS Safari  | macOS Chrome | iOS Safari |
  fail? `1` | pass? `2` | pass? `3` |

1. VO in macOS Safari is just barely usable in this example. (Looks like the groups are confusing things, because the original Scrollable Listbox Example works well). Up/down arrow keys move through the items, but VO doesn't read them. (maybe "activedescendant changed" events are not being sent?). You can still "interact with" the list, and use `ctrl+option+left/right arrow` to move through the options and `ctrl+option+space` to select, so I'm not sure it's a complete fail.

2. VO in macOS Chrome reads items and announces group switches with up/down arrow keys, but oddly, it counts the groups instead of the options, i.e. "Dog, Land, group (1 of 3)".
Typing End (fn+right arrow) announces group, i.e. "Owl, Air, group (3 of 3)", but Home (fn+left arrow) does not say group (just says "Cat").

3. Only tested basic swipe (to move through options) and double-tap (to select current option) in VO iOS. Can navigate to group name, but groups are not recognized as groups. Couldn't test keys like Home and End (no keyboard).

    macOS 10.15.1 Catalina
Safari 13.0.3
Chrome 78.0.3904.87
iOS 13.2

**Comment by carmacleod** [comment]
I noticed a minor functional problem while I was testing a11y.
If you type `End`, `Home`, `up arrow`, the list doesn't scroll up to show the Land category title.
You have to use `PageUp`.
Up arrow should probably also scroll to the top in this case.

If we don't want to hold up merging the example for this, I can open a new issue.

**Comment by mcking65** [comment]
@carmacleod

> I noticed a minor functional problem while I was testing a11y.
> If you type `End`, `Home`, `up arrow`, the list doesn't scroll up to show the Land category title.
> You have to use `PageUp`.
> Up arrow should probably also scroll to the top in this case.
> 
> If we don't want to hold up merging the example for this, I can open a new issue.

Carolyn, it would be really helpful if you would please open a separate issue for this and associate as with the listbox project, the 1.2 release 1 milestone, and tag as an example bug.

Many thanks for all the review work!!

**Comment by carmacleod** [comment]
@mcking65 
> Carolyn, it would be really helpful if you would please open a separate issue for this and associate as with the listbox project, the 1.2 release 1 milestone, and tag as an example bug.

Done. Opened https://github.com/w3c/aria-practices/issues/1249.

**Pull request merged**